### ISSUE TITLE
8250_sunxi.ko needs custom irq handler

### ISSUE THREAD

**Issue by amery** [linux-sunxi/linux-sunxi#28]
74270827595baecd329fd74f17f6b62c0738bcad made quite a few changes in the core 8250.ko, mostly debugging but there is an special irq handling (see `UART_USR`). Most of the patch can be forward ported without much troubles, but the following was done on a section removed by upstream in 4834d028978583dfe8e1fc19f1180ceb03d8dfb7 so we need to make our own irq handler like s3c64xx did in 88bb4ea14c72f89c5265029c1891e5eb6521cc0f.

``` diff
--- a/drivers/tty/serial/8250.c
+++ b/drivers/tty/serial/8250.c
@@ -1667,14 +1727,28 @@ static irqreturn_t serial8250_interrupt(int irq, void *dev_id)
                         * UART was busy. The interrupt must be cleared by reading
                         * the UART status register (USR) and the LCR re-written. */
                        unsigned int status;
-                       status = *(volatile u32 *)up->port.private_data;
+                       DEBUG_SERIAL(&up->port, "bus busy...\n");
+                       printk(">>> bus busy...\n");
+                       //status = *(volatile u32 *)up->port.private_data;
+                       status = serial_in(up, UART_USR);
                        serial_out(up, UART_LCR, up->lcr);

                        handled = 1;

                        end = NULL;
```

**Comment by amery** [comment]
maybe not... 8275a753705fae7a96d2606c01187d163804f506 removed the UPIO_DWAPB32 type part from the sunxi driver

**Comment by amery** [comment]
nope, arch/arm/mach-sun4i/devices.c defines the iotype of `debug_uart_platform_data` to `UPIO_DWAPB32`, allegedly to support busy busses.

**Comment by amery** [comment]
7d4008ebb1c971ce4baf57e45993690b0fa6d9f9 added a separated driver for DesignWare, which was the only legal user of `UPIO_DWAPB32`, and so this constant got removed in 4834d028978583dfe8e1fc19f1180ceb03d8dfb7

**Comment by amery** [comment]
second time I get bitten by this.

```
<3>>>> no handle, treat it handle over                                                                         
[    3.490000] >>> no handle, treat it handle over                                                             
<3>>>> no handle, treat it handle over                                                                         
[    3.490000] >>> no handle, treat it handle over                                                             
<3>>>> no handle, treat it handle over                                                                         
[    3.490000] >>> no handle, treat it handle over                                                             
<3>>>> no handle, treat it handle over                                                                         
[    3.490000] >>> no handle, treat it handle over 
```

**Comment by amery** [comment]
fix added to the `stage/sunxi-3.4` branch for further testing. 614fd293b370c88629c3c4437e1af5c0b013c087

**Comment by amery** [comment]
irq handler merged as 707f3413a38bf848da9422b6fa61b01dc1cd2a28 to `sunxi-3.4`. ticket closed.
